### PR TITLE
Promote adding headers by line to public API.

### PR DESCRIPTION
**Pull request by JakeWharton**

### CODE REVIEW

File: okhttp/src/main/java/com/squareup/okhttp/Headers.java [hunk 1]
```diff
@@ -172,7 +172,7 @@ public static Headers of(String... namesAndValues) {
     private final List<String> namesAndValues = new ArrayList<>(20);
 
     /** Add an header line containing a field name, a literal colon, and a value. */
-    Builder addLine(String line) {
+    public Builder add(String line) {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/Headers.java]: I think we want two APIs here.

addLine() is the awkward one that handles all the edge cases that deal with broken data, like lines `:foo: bar`. We support that to deal with old bugs in the response cache.

And the new one you're adding, `add` should be more strict. Maybe enforce a non-empty key?

**Comment by JakeWharton** [comment]
PTAL

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/HeadersTest.java [hunk 2]
```diff
@@ -145,6 +146,47 @@
...
+        .add("ping:  pong  ")
+        .add("kit:kat")
+        .add(" : baz")
+        .build();
+    assertEquals(Arrays.asList("bar"), headers.values("foo"));
+    assertEquals(Arrays.asList("bar"), headers.values("foo "));
+    assertEquals(Arrays.asList("bar"), headers.values(" foo"));
+    assertEquals(Arrays.asList("pong"), headers.values("ping"));
+    assertEquals(Arrays.asList("kat"), headers.values("kit"));
+    assertEquals(Arrays.asList("baz"), headers.values(" "));
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/HeadersTest.java]: this case is potentially sketchy. The spec is strict about what we should produce, and it says that ` ` is not a valid header name.
http://tools.ietf.org/html/rfc7230#section-3.2

I think we have to reject this input.

File: okhttp/src/main/java/com/squareup/okhttp/Headers.java [hunk 3]
```diff
@@ -171,8 +171,11 @@ public static Headers of(String... namesAndValues) {
   public static final class Builder {
     private final List<String> namesAndValues = new ArrayList<>(20);
 
-    /** Add an header line containing a field name, a literal colon, and a value. */
-    Builder addLine(String line) {
+    /**
+     * Add a header line without any validation. Only appropriate for headers from the remote peer
+     * or cache.
+     */
+    Builder addLenient(String line) {
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/Headers.java]: Thumbs up.

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/HeadersTest.java (hunk 2 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/HeadersTest.java]: Though this isn't what we're doing today. Ugh.

Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/HeadersTest.java]: Do we need this to be anything but super-strict? If not, let's trim.

**Comment by swankjesse** [comment]
LGTM

File: okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/HeadersTest.java (hunk 2 above)
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/internal/http/HeadersTest.java]: I'm for it. I had trim initially and remove it.

**Comment by JakeWharton** [comment]
Passed on one JDK. Failure looks to be a flake:
```
<<< FAILURE! - in com.squareup.okhttp.internal.http.URLConnectionTest
postFailsWithFixedLengthRequestForLargeRequest(com.squareup.okhttp.internal.http.URLConnectionTest)  Time elapsed: 0.022 sec  <<< ERROR!
java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:189)
	at java.net.SocketInputStream.read(SocketInputStream.java:121)
	at okio.Okio$2.read(Okio.java:136)
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:211)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:255)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:249)
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:196)
	at com.squareup.okhttp.internal.http.HttpConnection.readResponse(HttpConnection.java:188)
	at com.squareup.okhttp.internal.http.HttpTransport.readResponseHeaders(HttpTransport.java:80)
	at com.squareup.okhttp.internal.http.HttpEngine.readNetworkResponse(HttpEngine.java:791)
	at com.squareup.okhttp.internal.http.HttpEngine.readResponse(HttpEngine.java:678)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:431)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:376)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream(HttpURLConnectionImpl.java:227)
	at com.squareup.okhttp.internal.http.URLConnectionTest.assertContent(URLConnectionTest.java:3070)
	at com.squareup.okhttp.internal.http.URLConnectionTest.assertContent(URLConnectionTest.java:3074)
	at com.squareup.okhttp.internal.http.URLConnectionTest.reusedConnectionFailsWithPost(URLConnectionTest.java:2590)
	at com.squareup.okhttp.internal.http.URLConnectionTest.postFailsWithFixedLengthRequestForLargeRequest(URLConnectionTest.java:2568)
postFailsWithChunkedRequestForLargeRequest(com.squareup.okhttp.internal.http.URLConnectionTest)  Time elapsed: 0.012 sec  <<< ERROR!
java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:189)
	at java.net.SocketInputStream.read(SocketInputStream.java:121)
	at okio.Okio$2.read(Okio.java:136)
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:211)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:255)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:249)
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:196)
	at com.squareup.okhttp.internal.http.HttpConnection.readResponse(HttpConnection.java:188)
	at com.squareup.okhttp.internal.http.HttpTransport.readResponseHeaders(HttpTransport.java:80)
	at com.squareup.okhttp.internal.http.HttpEngine.readNetworkResponse(HttpEngine.java:791)
	at com.squareup.okhttp.internal.http.HttpEngine.readResponse(HttpEngine.java:678)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:431)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:376)
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream(HttpURLConnectionImpl.java:227)
	at com.squareup.okhttp.internal.http.URLConnectionTest.assertContent(URLConnectionTest.java:3070)
	at com.squareup.okhttp.internal.http.URLConnectionTest.assertContent(URLConnectionTest.java:3074)
	at com.squareup.okhttp.internal.http.URLConnectionTest.reusedConnectionFailsWithPost(URLConnectionTest.java:2590)
	at com.squareup.okhttp.internal.http.URLConnectionTest.postFailsWithChunkedRequestForLargeRequest(URLConnectionTest.java:2560)
```

**Comment by JakeWharton** [comment]
Wait now it's green on both.. Maybe that was an old build.

**Pull request merged**